### PR TITLE
[Subtitles] Avoid using hardcoded style ids

### PR DESCRIPTION
**Pull request by enen92**
## Description

This PR tries to improve a bit the way we handle libass styles for converted overlay types - by avoiding the usage of hardcoded style ids. Making assumptions about the presence (or absence) of a given default style id provided beforehand by libass is wrong as it might change between library versions (and strongly depends on the library internals). So this PR allocates a default kodi style id before ass_events are created so that the id is set on all events. When applying the style properties from the overlayrender, this style is then "enabled" and used when populating the struct.

@CastagnaIT looks like I can't add you as a reviewer. Can you please also take a look?

## Motivation and context

Should fix the issues with subtitles not being displayed in Ubuntu (at least it does for me).

## How has this been tested?

Runtime tested with native and converted subs

## What is the effect on users?
Should fix converted subtitles not displaying with different library versions

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [x] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

File: xbmc/cores/VideoPlayer/DVDSubtitles/SubtitlesAdapter.cpp [hunk 1]
```diff
@@ -27,7 +27,7 @@ CSubtitlesAdapter::~CSubtitlesAdapter()
 bool CSubtitlesAdapter::Initialize()
 {
   m_libass->SetSubtitleType(ADAPTED);
-  return m_libass->CreateTrack();
+  return m_libass->CreateTrack() && m_libass->CreateStyle();
```
Comment (howie-f) [inline xbmc/cores/VideoPlayer/DVDSubtitles/SubtitlesAdapter.cpp:30]: a heads up because i saw this in a different section: `&&` does short-circuit, so if `CreateTrack()` returns false -> `CreateStyle()` will not get executed. if that‘s not intended please use `&` …. or to make it more obvious:
```c
const bool trackCreated = …
const bool styleCreated = …
return trackCreated && styleCreated;

```

**Comment by wsnipex** [comment]
this still doesn't fix standard (embedded or external) srt subs.

> 2021-10-02 07:11:19.676 T:2317594    INFO <general>: CDVDSubtitlesLibass: Using libass version 1502000
> 2021-10-02 07:11:19.676 T:2317594    INFO <general>: CDVDSubtitlesLibass: Creating ASS library structure
> 2021-10-02 07:11:19.676 T:2317594    INFO <general>: CDVDSubtitlesLibass: Initializing ASS Renderer
> 2021-10-02 07:11:19.676 T:2317594   DEBUG <general>: CDVDSubtitlesLibass: [ass] libass API version: 0x1502000
> 2021-10-02 07:11:19.676 T:2317594   DEBUG <general>: CDVDSubtitlesLibass: [ass] libass source: tarball: 0.15.2
> 2021-10-02 07:11:19.677 T:2317594   DEBUG <general>: CDVDSubtitlesLibass: [ass] Shaper: FriBidi 1.0.8 (SIMPLE) HarfBuzz-ng 2.6.4 (COMPLEX)
> 2021-10-02 07:11:19.677 T:2317594    INFO <general>: CDVDSubtitlesLibass: Initializing ASS library font settings
> 2021-10-02 07:11:19.677 T:2317594   DEBUG <general>: Created subtitles parser: SubRip Subtitle Parser
>

File: xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp [hunk 2]
```diff
@@ -258,15 +271,7 @@ void CDVDSubtitlesLibass::ApplyStyle(style subStyle, renderOpts opts)
...
-      if (m_currentStyleId != ASS_NO_ID)
-      {
-        // Delete current style
-        ass_free_style(m_track, m_currentStyleId);
-        m_track->n_styles -= 1;
-      }
-      m_currentStyleId = ass_alloc_style(m_track);
-      m_track->default_style = m_currentStyleId;
-      style = m_track->styles + m_currentStyleId;
+      style = &m_track->styles[m_currentDefaultStyleId];
```
Comment (CastagnaIT) [inline xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:274]: now that the override ASS settings can be changed only when not in playing,
reuse always same style, how did you, should be fine

File: xbmc/cores/VideoPlayer/DVDSubtitles/SubtitlesAdapter.cpp (hunk 1 above)
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/DVDSubtitles/SubtitlesAdapter.cpp:30]: CreateStyle depends exclusively on CreateTrack i think go well with `&&`

**Comment by CastagnaIT** [comment]
@wsnipex i think @enen92 have tested on old ubuntu 18 version not the last Ubuntu, @enen92 can you specify on what versions?

with last ubuntu + Kodi 19 ASS subtitles not works, this means that the problem was already there before, therefore the problem is not the changes made with the rework PR, but probable on kodi rendering

**Comment by enen92** [comment]
Yeah I think we have 2 different issues at play. This was on ubuntu18.04 with the nightly ppa and libass 0.14. ASS style subs worked but converted subs did not show because of the hardcoded style IDs. The other issue is probably something related to the rendering pipeline, I'll try to setup a similar environment (Ubuntu 20.04?) but I think that's likely out of my league :)

File: xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp (hunk 2 above)
Comment (enen92) [reply xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:274]: Can you rephrase please? I did not understand. You mean this is wrong or right? Is forcing ass styles supposed to work during runtime playback and this breaks the functionality?

File: xbmc/cores/VideoPlayer/DVDSubtitles/SubtitlesAdapter.cpp (hunk 1 above)
Comment (enen92) [reply xbmc/cores/VideoPlayer/DVDSubtitles/SubtitlesAdapter.cpp:30]: Yeah this is intentional. I'll add logging prior to all early returns both in createtrack and createstyle

File: xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp (hunk 2 above)
Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:274]: my fear was that ass overrides settings might cause inconsistencies with style properties,
in short now i agree with this change

Comment (enen92) [reply xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:274]: I see, runtime tested and worked fine. Found a "bug" though (which might be in the settings internal implementation though):
- start playing a file and go to the settings window
- check that the ass override setting is disabled as it should
- hit "x" to stop the video while in this window
- check that the ass override setting is disabled which is wrong as nothing is playing
image

- go back to another window and back to the settings window
- check that the setting is now enabled (correctly because nothing is playing)

Not a regression for sure, just something that likely was not tested found before

Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/DVDSubtitles/DVDSubtitlesLibass.cpp:274]: maybe somewhere is needed to implement a way to receive callbacks like player status to update the list of settings, if it is curently displayed
probably is not implemented because my addition to check playback status is new

**Comment by wsnipex** [comment]
> @wsnipex i think @enen92 have tested on old ubuntu 18 version not the last Ubuntu, @enen92 can you specify on what versions?
> 
> with last ubuntu + Kodi 19 ASS subtitles not works, this means that the problem was already there before, therefore the problem is not the changes made with the rework PR, but probable on kodi rendering

ass subs work, but not standard srt. And it definitely is the removal of legacy rendering.

**Comment by enen92** [comment]
Freshly compiled kodi on ubuntu 20.04 from the branch of this PR and everything works fine for me, either ASS based subs or adapted ones:

https://imgur.com/a/ZngmQfp

What am I missing?

**Comment by wsnipex** [comment]
what is an "adapted" sub? As said, ASS subs work, but not plain standard SRT.

**Comment by enen92** [comment]
I mean all text subtitles that are not ASS/SSA like srt, sami, etc. Now they are converted to ass and rendered via libass (so they end up being ASS subtitles as well).
I don't seem to be able to reproduce the issue, my srt subtitles render just fine

**Comment by CastagnaIT** [comment]
i have tried install Ubuntu 21 in to another pc and then install last kodi nightly (20211001-45b6c61258)
tested ASS subtitles and SRT subtitles mkv embedded and external files and all works as expected (my previous tests on VM was not affidable)

is possible some opengl compatibility problem to be resolved in kodi with his graphics card (Mesa Intel(R) UHD Graphics)?
i have no other idea
maybe @wsnipex can try to do a clean install of ubuntu on another hdd to make clean test (if it has not already done so...)

**Comment by enen92** [comment]
jenkins build this please

**Comment by enen92** [comment]
@lrusak since you tested this and fixed your issue, can I get a green mark for merging?

**Review by howie-f — APPROVED** [review]
i don't see anything wrong with this
